src/backend/catalog/storage_pending_deletes_redo.c

```diff
@@ -0,0 +1,324 @@
+/*-------------------------------------------------------------------------
+ *
+ * storage_pending_deletes_redo.c
+ *	  code to support processing of pending deletes (orphaned files) in WAL
+ *
+ * Copyright (c) 2025 Greengage Community
+ *
+ *	  src/backend/catalog/storage_pending_deletes_redo.c
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+
+#include "access/clog.h"
+#include "access/transam.h"
+#include "catalog/storage_pending_deletes_redo.h"
+#include "miscadmin.h"
+#include "storage/md.h"
+#include "utils/elog.h"
+#include "utils/guc.h"
+#include "utils/hsearch.h"
+
+/*
+ * HTAB entry for pending deletes for the given xid.
+ */
+typedef struct PendingDeleteHtabNode
+{
+	TransactionId xid;
+	List	   *relnode_list;	/* list of RelFileNodePendingDelete */
+}	PendingDeleteHtabNode;
+
+/*
+ * Hash table for pending deletes relfilenodes for a given xid.
+ */
+static HTAB *pendingDeletesRedo = NULL;
+
+/*
+ * This function inserts XLOG_PENDING_DELETE record into WAL.
+ */
+void
+PdlXLogInsert()
+{
+	Size		size = 0;
+
+	if (IsBootstrapProcessingMode() || !gp_track_pending_delete)
+		return;
+
+	PendingRelXactDeleteArray *arr = PdlXLogShmemDump(&size);
+
+	if (size > 0 && arr != NULL)
+	{
+		XLogRecPtr	rec;
+		XLogRecData rdata =
+		{
+			.buffer = InvalidBuffer,
+			.data = (char *) arr,
+			.len = size,
+			.next = NULL,
+			.buffer_std = false
+		};
+
+		rec = XLogInsert(RM_XLOG_ID, XLOG_PENDING_DELETE, &rdata);
+
+		XLogFlush(rec);
+
+		elog(DEBUG1, "Pending delete XLog record inserted");
+
+		pfree(arr);
+	}
+}
+
+/*
+ * This function adds pending delete node to a pendingDeletesRedo hash-table
+ * during WAL redo processing.
+ */
+void
+PdlRedoAdd(PendingRelXactDelete * pd)
+{
+	Assert(pd);
+
+	if (IsBootstrapProcessingMode() ||
+		(pd->xid == InvalidTransactionId) ||
+		!gp_track_pending_delete)
+		return;
+
+	if (NULL == pendingDeletesRedo)
+	{
+		HASHCTL		info =
+		{
+			.keysize = sizeof(TransactionId),
+			.entrysize = sizeof(PendingDeleteHtabNode)
+		};
+
+		pendingDeletesRedo = hash_create("pendingDeletesRedo hash",
+										 32,
+										 &info,
+										 HASH_ELEM);
+	}
+
+	bool		found = false;
+
+	PendingDeleteHtabNode *entry =
+	(PendingDeleteHtabNode *) hash_search(pendingDeletesRedo, &pd->xid, HASH_ENTER, &found);
```

Comment:
This line has more than 80 symbols (92). There are other lines with more than 80 symbols.